experimental/incremental/task.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package incremental
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Task represents a query that is currently being executed.
+//
+// Values of type Task are passed to [Query]. The main use of a Task is to
+// be passed to [Resolve] to resolve dependencies.
+type Task struct {
+	// We need all of the contexts for a call to [Run] to be the same, so to
+	// avoid user implementations of Query making this mistake (or inserting
+	// inappropriate timeouts), we pass the context as part of the task context.
+	ctx    context.Context //nolint:containedctx
+	exec   *Executor
+	task   *task
+	result *result
+}
+
+// Error adds errors to the current query, which will be propagated to all
+// queries which depend on it.
+//
+// This will not cause the query to fail; see [Task.Fail] for that.
+func (t *Task) Error(errs ...error) {
+	t.result.Errors = append(t.result.Errors, errs...)
+}
+
+// Fail marks the current query as failed. This will cause all queries that
+// depend on this one to also fail.
+//
+// This function will not return: it will cause a goroutine which calls it to
+// exit. See [Resolve] for notes on what this entails.
+func (t *Task) Fail(errs ...error) {
+	t.Error(errs...)
+	t.result.Failed = true
+	close(t.result.done)
+	runtime.Goexit()
+}
+
+// Resolve executes a set of queries in parallel. Each query is run on its own
+// goroutine.
+//
+// If any of the queries fails, or if the [context.Context] passed to the
+// [Run] call that spawned the [Task] is cancelled, this function calls
+// [runtime.Goexit]. This is not something callers should be concerned about,
+// because every call to a Query execution function is done in its own
+// goroutine, and only occurs when the return value of [Query].Execute will be
+// discarded.
+//
+// Errors for each result are only those that occurred as a direct result of
+// query execution, and *does not* contain that query's transitive errors. This
+// is unlike the behavior of [Run].
+func Resolve[T any](caller Task, queries ...Query[T]) []Result[T] {
+	results := make([]Result[T], len(queries))
+	deps := make([]*task, len(queries))
+	wg := semaphore.NewWeighted(int64(len(queries)))
+
+	for i, q := range queries {
+		i := i
+		deps[i] = caller.exec.getTask(q.URL())
```

Comment:
We should also clearly document that `Query.Execute` should never call `Run`, only `Resolve`. (Otherwise, we'd lose the trail of callers and not be able to detect cycles.)

Reply:
Implemented cycle tracking. There's even a test!

In the process I found a deadlock due to resource starvation. There's some fussiness in Resolve to drop the hold on exec.sema in the right place to avoid starvation.